spec/support/fake_io.rb

```diff
@@ -1,29 +1,31 @@
+# frozen_string_literal: true
 
-class FakeJSONIO
-  def close
+class FakeIO
+  attr_reader :received_messages
+
+  def initialize
+    @msg = nil
+    @received_messages = []
   end
 
+  def close; end
+
   def write(msg)
     @msg = msg
+    @received_messages << @msg
   end
 
   def received_message
-    @msg && JSON.parse(@msg)
+    @msg
   end
 end
 
-class FakeIO
-  def initialize
-    @msg = nil
-  end
-  def close
-  end
-
-  def write(msg)
-    @msg = msg
+class FakeJSONIO < FakeIO
+  def received_message
+    @msg && Oj.load(@msg)
   end
 
-  def received_message
-    @msg
+  def received_messages
+    @received_messages.map { |m| Oj.load(m) }
   end
 end
```

Comment:
the diff is a bit hard to comprehend but in a nutshell,
1) made `FakeJSONIO` inherit from `FakeIO`
2) added a `received_messages` (plural) method to keep track of all messages. This way we can test areas where the logger is called multiple times.